### DISCUSSION TITLE
Possible to see transformed time series data?

### DISCUSSION THREAD

**Discussion by stikket** [winedarksea/AutoTS#232] [Q&A]
Hi everyone, great package. 

Is it possible to see the transformed version of training data for timeseries? Is this feature available already in the package?

**Comment by winedarksea** [comment]
I assume you are looking for what the data looks like after the preprocessing transforms are applied, as the underlying model would see the data?

Ensembles can mess things up a bit because they mix together separate transformer+model groups, but generally:

```python
import json
from autots import GeneralTransformer, AutoTS

model = AutoTS(...your_args...).fit()

# Ensembles don't have one overall transformer, so we can't plot that
# but for horizontal with no style 1 ensembles we can still view the best non ensemble model transformer
if model.best_model_ensemble == 2:
    params = json.loads(model.best_model_non_horizontal['TransformationParameters'].iloc[0])
else:
    params = model.best_model_transformation_params
    params = model.best_model_transformation_params
# if this is an empty dict, this means that best model and best non horizontal are both an Ensemble
# you could loop through the Ensemble params and plot each transformer, in that case, however, not shown
print(f"Transformer params are {params}")

transformer = GeneralTransformer(**params)
transformed_df = transformer.fit_transform(model.df_wide_numeric)

column = model.df_wide_numeric.columns[0]
plot_df = model.df_wide_numeric[column].to_frame().merge(transformed_df[column].to_frame().rename(columns=lambda x: "transformed_" + str(x)[0:50]), right_index=True, left_index=True)
plot_df.plot() # scale is often very different between two, because scaling
plot_df.iloc[:, -1].plot() # just the transformed
# plt.show()
```

**Reply by stikket** [reply]
Thank you for giving a very concise answer. 

A few follow up question:
 If I have a single column to forecast and have multiple transforms, the transformed_df would still have a single column? I was assuming it would add columns (or is it the time series model itself that creates additional features? i.e. ARIMA and it would just use the transformed column as input?)

**Reply by winedarksea** [reply]
So there are two different things which are both important:
1. The preprocessing transformers. These work in 'layers' to add and remove various aspects to the data around the primary model
2. The primary model, which in the case of ~Regression style models and some others, will have an X and a Y feature set which do have columns of features created from the input series.
My previous answer focused on point 1.

So, details:

- Most transformers don't add columns. They instead work in layers adding and removing but mostly keeping the original data shape at the end. The exceptions are `CenterSplit` and anything with a macro_micro arg which do double the number of columns by splitting the data
- If you are looking for models that provide an X and a Y feature set for other machine learning, the easiest to extract that from are the `DatepartRegression`, `MultivariateRegression`, and `WindowRegression`. You can optimize these with AutoTS, then use both the transformers and model parameters to extract a feature set. I can provide code for this if you need help with that.
- PreprocessingRegression (still kinda beta) is a model that mixes the preprocessing as the X feature set, so it is an exception.
